internal/lifecycle/lifecycle.go

```diff
@@ -0,0 +1,108 @@
+// Package lifecycle provides a streamlined way to manage the application
+// lifecycle, queueing up runs and waiting for them to finish, returning all
+// errors that occurred during the runs.
+package lifecycle
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ErrTimeoutReached is returned if the lifecycle times out.
+var ErrTimeoutReached = errors.New("timeout reached")
+
+// Lifecycle manages the application lifecycle.
+type Lifecycle struct {
+	timeout time.Duration
+	counter sync.WaitGroup
+
+	firstRun sync.Once
+	quitMu   sync.RWMutex
+	appMu    sync.Mutex
+
+	errors []error
+
+	exitRequested bool
+	quitted       chan struct{}
+	started       chan struct{}
+}
+
+// New returns a new Lifecycle with the given timeout.
+func New(timeout time.Duration) *Lifecycle {
+	return &Lifecycle{
+		timeout: timeout,
+		quitted: make(chan struct{}),
```

Comment:
No really thought about it fully yet: but would it be possible to merged started/quitted by using the `nil` value smartly? (Remember that we can check for a nil channel, listening on it will be undefinitively)